### PR TITLE
Feature/gh 10920 enable strict mode

### LINKED ISSUE #10920: Enable strict mode for common entry point of product-configurator library
Have
"strict": true
in tsconfig.lib.json and fix the issues

### PR DESCRIPTION
**Pull request by ChristophHi**

### CODE REVIEW

File: feature-libs/product-configurator/common/components/configurator-issues-notification/configurator-issues-notification.component.html [hunk 1]
```diff
@@ -8,11 +8,14 @@
...
-        *ngIf="orderEntry?.product?.configurable"
+        *ngIf="
+          orderEntry?.product?.configurable && quantityControl$
+            | async as quantityControl
+        "
         [cartEntry]="orderEntry"
         [readOnly]="readonly$ | async"
         [msgBanner]="true"
-        [disabled]="(quantityControl$ | async).disabled"
+        [disabled]="quantityControl.disabled"
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/common/components/configurator-issues-notification/configurator-issues-notification.component.html:18]: Would be good if you could explain this change to the team, at least i known not enough about the async pipe construct to even understand the issue.

File: feature-libs/product-configurator/common/components/configurator-cart-entry-info/configurator-cart-entry-info.component.ts [hunk 2]
```diff
@@ -31,9 +31,11 @@ export class ConfiguratorCartEntryInfoComponent {
...
-    return (
-      item?.configurationInfos?.length > 0 &&
-      item?.configurationInfos[0]?.status !== 'NONE'
-    );
+    const configurationInfos = item?.configurationInfos;
+
+    return configurationInfos
+      ? configurationInfos.length > 0 &&
+          configurationInfos[0]?.status !== 'NONE'
+      : false;
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/common/components/configurator-cart-entry-info/configurator-cart-entry-info.component.ts:39]: The two added tests are already green with old version of the code, so this isn't fixing anything?
Or is this more about code-style, as the new version is a bit more explicit and maybe easier to understand. (no comparing with undefined)

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
In general most things look valid to me. It seems they address unintended returns of undefined, which is good. 👍 
only the first issue surprised me, as it is not clear to me what should be solved/fixed here. See my comment.

File: feature-libs/product-configurator/common/components/configure-product/configure-product.component.spec.ts [hunk 3]
```diff
@@ -41,8 +51,22 @@ let currentProductService: CurrentProductService;
...
+    TestBed.configureTestingModule({
+      imports: [I18nTestingModule, RouterTestingModule],
+      declarations: [ConfigureProductComponent, MockUrlPipe],
+      providers: [
+        {
+          provide: CurrentProductService,
+          useClass: MockCurrentProductServiceReturnsNull,
+        },
+      ],
+    }).compileComponents();
```
Comment (Platonn) [inline feature-libs/product-configurator/common/components/configure-product/configure-product.component.spec.ts:68]: I agree the setup of TestBed deserves an util function. But mocking method properties should not be done there IMHO. We don't need a separate mock service class too. Can we instead simply add `spyOn` locally in the test that checks against null? 

```typescript
it('.....', () => {
  const service = TestBed.inject(CurrentProductService);
  spyOn(service, 'getProduct).and.return(of(null))`

  /* ... */
})
```

File: feature-libs/product-configurator/common/components/configurator-issues-notification/configurator-issues-notification.component.html (hunk 1 above)
Comment (ChristophHi) [reply feature-libs/product-configurator/common/components/configurator-issues-notification/configurator-issues-notification.component.html:18]: Ok, will do

File: feature-libs/product-configurator/common/components/configurator-cart-entry-info/configurator-cart-entry-info.component.ts (hunk 2 above)
Comment (ChristophHi) [reply feature-libs/product-configurator/common/components/configurator-cart-entry-info/configurator-cart-entry-info.component.ts:39]: This fixes the strict mode compliance, but your are right, the old version also lets the tests pass (because item?.configurationInfos?.length can be compared to 0 even if configurationInfos is undefined. But besides strict mode compliance that is hard to understand, and one even doesn't notice immediately why the code lets the test pass)

**Review by Uli-Tiger — APPROVED** [review]

File: feature-libs/product-configurator/common/components/configure-product/configure-product.component.ts [hunk 4]
```diff
@@ -14,13 +15,18 @@ import { ConfiguratorProductScope } from '../../core/model/configurator-product-
...
     : this.currentProductService
     ? this.currentProductService.getProduct(
         ConfiguratorProductScope.CONFIGURATOR
       )
-    : of(null);
+    : of(null)
+  ).pipe(
+    //needed because also currentProductService might return null
+    map((product) => (product ? product : this.nonConfigurable))
+  );
```
Comment (Platonn) [inline feature-libs/product-configurator/common/components/configure-product/configure-product.component.ts:29]: Did we consider having `Observable<Product | null>` instead? Side note: I believe the CurrentProductService.getProduct() should have had return type `Observable<Product | null>` anyway.

Returning an incomplete product object looks not natural to me and can be misleading (it looks like the product exists, but doesn't have nearly any properties).

File: feature-libs/product-configurator/common/components/configure-product/configure-product.component.spec.ts [hunk 5]
```diff
@@ -122,6 +146,14 @@ describe('ConfigureProductComponent', () => {
     });
   });
 
+  it('should emit non-configurable dummy in case it was launched with product service which emits null', (done) => {
+    setupWithCurrentProductService(true, true);
```
Comment (Platonn) [inline feature-libs/product-configurator/common/components/configure-product/configure-product.component.spec.ts:150]: Nitpick: When we really need the util function to setup a test, it would be good if arguments are readable (i.e. change unnamed arguments to an object with named properties, `setupWithCurrentProductService(true, true)` -> `setupTest({productService: CurrentProductService, returnsNull: true}`).

File: feature-libs/product-configurator/common/components/service/configurator-router-extractor.service.ts [hunk 6]
```diff
@@ -36,11 +36,11 @@ export class ConfiguratorRouterExtractorService {
...
             routingData.state.queryParams?.resolveIssues === 'true',
           forceReload: routingData.state?.queryParams?.forceReload === 'true',
-          pageType: routingData.state.semanticRoute.includes(
-            this.ROUTE_FRAGMENT_OVERVIEW
-          )
-            ? ConfiguratorRouter.PageType.OVERVIEW
-            : ConfiguratorRouter.PageType.CONFIGURATION,
+          pageType:
+            semanticRoute &&
+            semanticRoute.includes(this.ROUTE_FRAGMENT_OVERVIEW)
```
Comment (Platonn) [inline feature-libs/product-configurator/common/components/service/configurator-router-extractor.service.ts:41]: Nitpick: For conciseness, did we consider `routingData.state?.semanticRoute?.includes(...)`?

File: feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts [hunk 7]
```diff
@@ -90,7 +95,10 @@ export class CommonConfiguratorUtilsService {
     let numberOfIssues = 0;
     cartItem?.statusSummaryList?.forEach((statusSummary) => {
       if (statusSummary.status === OrderEntryStatus.Error) {
-        numberOfIssues = statusSummary.numberOfIssues;
+        const numberOfIssuesFromStatus = statusSummary.numberOfIssues;
+        numberOfIssues = numberOfIssuesFromStatus
+          ? numberOfIssuesFromStatus
+          : 0;
```
Comment (Platonn) [inline feature-libs/product-configurator/common/shared/utils/common-configurator-utils.service.ts:101]: Nitpick: For conciseness, did we consider `numberOfIssues = statusSummary?.numberOfIssues ?? 0`  ?
